src/librustc_typeck/diagnostics.rs

```diff
@@ -138,6 +209,110 @@ enum Empty {}
 ```
 "##,
 
+E0106: r##"
+This error indicates that a lifetime is missing from a type. If it is an error
+inside a function signature, the problem may be with failing to adhere to the
+lifetime elision rules (see below).
+
+Here are some simple examples of where you'll run into this error:
+
+```
+struct Foo { x: &bool }        // error
+struct Foo<'a> { x: &'a bool } // correct
+
+enum Bar { A(u8), B(&bool), }        // error
+enum Bar<'a> { A(u8), B(&'a bool), } // correct
+
+type MyStr = &str;        // error
+type MyStr<'a> = &'a str; //correct
+
+```
+
+Lifetime elision is a special, limited kind of inference for lifetimes in
+function signatures which allows you to leave out lifetimes in certain cases.
+For example, the lifetimes on parameter in the following function signatures
+have been left out, but they still compile successfully:
+
+```
+fn foo(x: &str) { }
+
+fn bar(x: &str, y: &str) { }
+
+fn baz(x: &str) -> &str { x }
+```
+
+To explain the lifetime elision rules, we need to first discuss some background.
```

Comment:
I wonder if this may not be the best place to talk about lifetime elision? These are definitely great docs, but perhaps they could be placed in a lifetime elision chapter of the book instead? (and then this would link to that chapter)